### ISSUE TITLE
Because of Chinese characters colnames,read_sav give an error:"Invalid file, or file has unsupported features"

### ISSUE THREAD

**Issue by ccccfys** [tidyverse/haven#467]
I use haven::read_sav() to read a spss file,when the file haven Chinese colnames and the colnames have more than 3 Chinese characters,It will read failed and give an error：“Invalid file, or file has unsupported features.”  The following code repeats the problem。

This problem is solved by changing the column names to English.

（使用haven包中的read_sav函数读取spss文件时，如果其中的变量名称包含3个以及3个以上中文字符，则会读取失败，解决办法是把缩减名称至两个中文字符，或把变量名称改成英文。下方代码重现了这个问题。）

```r
> library(haven)

> info <- data.frame(id = 1:2,
+                    sex = factor(c('男','女')),
+                    age = c(25,34))

> # write with english colnames
> colnames(info)
[1] "id"  "sex" "age"

> write_sav(info,'info_01.sav');read_sav('info_01.sav')
# A tibble: 2 x 3
     id       sex   age
  <dbl> <dbl+lbl> <dbl>
1     1    1 [男]    25
2     2    2 [女]    34

> # write with Chinese colnames(2 Chinese characters) 
> colnames(info)[2] <- "性别";colnames(info)
[1] "id"   "性别" "age" 

> write_sav(info,'info_02.sav');read_sav('info_02.sav')
# A tibble: 2 x 3
     id      性别   age
  <dbl> <dbl+lbl> <dbl>
1     1    1 [男]    25
2     2    2 [女]    34

> # write with Chinese colnames(3 Chinese characters) 
> colnames(info)[2] <- "是性别";colnames(info)
[1] "id"     "是性别" "age"   

> write_sav(info,'info_03.sav');read_sav('info_03.sav')
Error parsing string "ID=id	是性�=是性别	AGE=age" around byte #14/32, character =
Error in df_parse_sav_file(spec, encoding, user_na) : 
  Failed to parse /Volumes/File/RInAction/data/chapter07/info_03.sav: Invalid file, or file has unsupported features.
```

**Comment by hadley** [comment]
Could you please rework your reproducible example to use the [reprex package](http://reprex.tidyverse.org) ? That makes it easier to see both the input and the output, formatted in such a way that I can easily re-run in a local session.

**Comment by ccccfys** [comment]
```r
library(haven)

info <- data.frame(id = 1:2,
                   sex = factor(c('男','女')),
                   age = c(25,34))

# write with english colnames
colnames(info)
#> [1] "id"  "sex" "age"
write_sav(info,'info_01.sav');read_sav('info_01.sav')
#> # A tibble: 2 x 3
#>      id       sex   age
#>   <dbl> <dbl+lbl> <dbl>
#> 1     1    1 [男]    25
#> 2     2    2 [女]    34

# write with Chinese colnames(2 Chinese characters) 
colnames(info)[2] <- "性别";colnames(info)
#> [1] "id"   "性别" "age"
write_sav(info,'info_02.sav');read_sav('info_02.sav')
#> # A tibble: 2 x 3
#>      id      性别   age
#>   <dbl> <dbl+lbl> <dbl>
#> 1     1    1 [男]    25
#> 2     2    2 [女]    34

# write with Chinese colnames(3 Chinese characters) 
colnames(info)[2] <- "是性别";colnames(info)
#> [1] "id"     "是性别" "age"
write_sav(info,'info_03.sav')   # write is ok

read_sav('info_03.sav')         # read does't work
#> Error parsing string "ID=id	是性�=是性别	AGE=age" around byte #14/32, character =
#> Error in df_parse_sav_file(spec, encoding, user_na, cols_skip, n_max) : 
#> Failed to parse /Users/songshu/info_03.sav: Invalid file, or file has unsupported features.

```

**Comment by ccccfys** [comment]
> Could you please rework your reproducible example to use the [reprex package](http://reprex.tidyverse.org) ? That makes it easier to see both the input and the output, formatted in such a way that I can easily re-run in a local session.

```r
library(haven)

info <- data.frame(id = 1:2,
                   sex = factor(c('男','女')),
                   age = c(25,34))

# write with english colnames
colnames(info)
#> [1] "id"  "sex" "age"
write_sav(info,'info_01.sav');read_sav('info_01.sav')
#> # A tibble: 2 x 3
#>      id       sex   age
#>   <dbl> <dbl+lbl> <dbl>
#> 1     1    1 [男]    25
#> 2     2    2 [女]    34

# write with Chinese colnames(2 Chinese characters) 
colnames(info)[2] <- "性别";colnames(info)
#> [1] "id"   "性别" "age"
write_sav(info,'info_02.sav');read_sav('info_02.sav')
#> # A tibble: 2 x 3
#>      id      性别   age
#>   <dbl> <dbl+lbl> <dbl>
#> 1     1    1 [男]    25
#> 2     2    2 [女]    34

# write with Chinese colnames(3 Chinese characters) 
colnames(info)[2] <- "是性别";colnames(info)
#> [1] "id"     "是性别" "age"
write_sav(info,'info_03.sav')   # write is ok

read_sav('info_03.sav')         # read does't work
#> Error parsing string "ID=id	是性�=是性别	AGE=age" around byte #14/32, character =
#> Error in df_parse_sav_file(spec, encoding, user_na, cols_skip, n_max) : 
#> Failed to parse /Users/songshu/info_03.sav: Invalid file, or file has unsupported features.
```

**Comment by hadley** [comment]
Somewhat more minimal reprex:

```R
library(haven)

df <- data.frame(x = factor(c('男','女')))
names(df) <- "是性别"

path <- tempfile()
write_sav(df, path)
read_sav(path)
```

Possibly suspiciously this fails when I attempt to use reprex with `input string 1 is invalid UTF-8`

**Comment by evanmiller** [comment]
Probably related to https://github.com/WizardMac/ReadStat/issues/206

**Comment by hadley** [comment]
Now works, thanks to readstat update.

**Comment by deschen1** [comment]
I get the same error when redaing in an SPSS file with haven. So the bug seems not to be fixed entirely. In my case, the colnames are perfectly fine. Our suspicion is that it is due to potentially longer non-Western text responses in string vars in SPSS where the byte handling is either buggy in SPSS or ReadStat does not yet mirror SPSS's behaviour.

not sure, but could be related to this one here: https://github.com/tidyverse/haven/issues/615

If not, I'd be glad about reopening (although I can't provide a reproducible example since opening the file in SPSS and saving it already solves the issue + the file i'm talking about can't be shared publicly).

**Comment by gorcha** [comment]
Hi @deschen1, unfortunately without a reprex/some example data we can't investigate this further.
Was the SPSS file that you're seeing the issue with produced in haven or is it from another source?

**Comment by deschen1** [comment]
The file was created outside of haven. My hope is that this is really the same thing as in #615 as mentioned above. I feel this byte-handling thing causes many errors currently. I understand taht you can't do much without an example. So I put my hope into the fix for #615 currently and will report if this indeed solves the issue.

**Comment by gorcha** [comment]
No worries, there's a fix for #615 that's almost finalised over at ReadStat so hopefully this will be sorted out soon!
